servicenow/stackstate_checks/servicenow/servicenow.py

```diff
@@ -167,73 +199,181 @@ def _process_components(self, components, instance_config):
+        return self._get_json(url, instance_info.timeout, auth)
+
+    def _sanitize_response(self, cr_list):
+        sanitized_crs = []
+        for cr in cr_list:
+            sanitized_crs.append(self.filter_empty_metadata(cr))
+        return sanitized_crs
+
+    def _process_change_requests(self, instance_info):
+        state = self.persistent_state.get_state(self.persistent_instance)
+        latest_sys_updated_on = datetime.datetime.strptime(state.get(instance_info.url), TIME_FORMAT)
+        response = self._collect_change_requests(instance_info, latest_sys_updated_on)
+        sanitized_result = self._sanitize_response(response['result'])
+        crs_persisted_state = state[self.cr_persistence_key]
+        for cr in sanitized_result:
+            try:
+                change_request = ChangeRequest(cr, strict=False)
+                change_request.validate()
+            except DataError as e:
+                self.log.info('%s - DataError: %s. This CR is skipped.', cr.get('number'), e)
+                continue
+            if change_request.cmdb_ci:
+                if change_request.sys_updated_on > latest_sys_updated_on:
+                    latest_sys_updated_on = change_request.sys_updated_on
+                    state[instance_info.url] = str(latest_sys_updated_on)
+                old_state = crs_persisted_state.get(change_request.number)
+                if old_state is None or old_state != change_request.state:
+                    self._create_event_from_change_request(change_request, instance_info)
+                    crs_persisted_state[change_request.number] = change_request.state
+        self.persistent_state.set_state(self.persistent_instance, state)
+
+    def _load_state(self, instance_info):
+        """
+        Persistent State structure:
+        {
+            url: timestamp,
+            url/change_requests: {
+                change_request_number: state
+            }
+        }
+        timestamp is the date/time of the last sys_updated_on from processed CRs
+
+        :param instance_info: current instance check is processing
+        :return: None
+        """
+        # TODO switch to schematics model
+        self.persistent_state = PersistentState()
+        start_dt = datetime.datetime.now() - datetime.timedelta(days=CRS_OLDEST_UPDATE_IN_NUMBER_OF_DAYS)
+        try:
+            self.persistent_state.get_state(self.persistent_instance)
+        except StateReadException:
+            self.log.info('First run! Creating new persistent state.')
+            empty_state = {
+                instance_info.url: start_dt.strftime(TIME_FORMAT),
+                self.cr_persistence_key: {}
+            }
+            self.persistent_state.set_state(self.persistent_instance, data=empty_state)
+
+    def _create_event_from_change_request(self, change_request, instance_info):
+        cmdb_ci = change_request.cmdb_ci
+        identifiers = []
+        external_id = cmdb_ci['link'].split('/').pop()
```

Comment:
If you are using `ALL` value for `sysparam_display` as discussed yesterday then you will not get `link` i think.

Reply:
That's true. Currently it gets me just `display_values` not `values`. I can set it to `ALL` and get them both but I will need to change `ChangeRequest(Model):` types and `_create_event_from_change_request` method that gets values from that model.